### DISCUSSION TITLE
FastAPI cursor based pagination example

### DISCUSSION THREAD

**Discussion by shellking4** [fastapi/fastapi#8515] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

def get_all_with_pagination(*, dB: Session = Depends(database.get_db), page_number: int=1, page_size: int = 70):
  return engine_service.get_with_pagination(
 dB, page_number, page_size
)
```

### Description

I would like to have an example of how to implement a cursor based pagination in FastAPI. Any help would be very welcome. I managed to implement offset based pagination like the one shown in the docs but I read on the internet that offset based paginations are inefficient. Then I would like to see how to do cursor based ones in FastAPI + Sqlalchemy + PostgreSQL. The frontend is using datatable.net, how to put everything together to have a full working pagination system. Thank you for your help in advance

### Operating System

Linux

### FastAPI Version

0.75.0

### Python Version

3.10.4

**Comment by JarroVGIT** [comment, marked as answer]
This is not really an FastAPI issue at all but rather a general guidance on how to implement a cursor based pagination system for a database. Some general guidance:

- Read this [https://medium.com/swlh/how-to-implement-cursor-pagination-like-a-pro-513140b65f32](this) article on what queries you would expect from a cursor based pagination system
- Rethink the returned value of your API (e.g. it is the page itself, plus a prev_ and next_ value)
- Make sure you have a unique, sorted column in your database (this is a requirement!)
- Rethink the parameter(s) that your api receives; does it receive only 1 param in the format of `next__{id}` or `prev__{id}` or does it have multiple?
- If you have very specific questions on problems that you run into, you might be better helped on Stackoverflow.

In general, cursor based pagination can greatly speed up **certain** use cases, but offset pagination is absolutely not dead. 

Good

**Comment by acidjunk** [comment]
We use the offset based pagination: 

https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/490c554e23343eec0736b06e59b2108fdd057fdc/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/crud/base.py#L29

Satisfied about performance etc..
(even extended it with filters/search)

**Comment by shellking4** [comment]
> This is not really an FastAPI issue at all but rather a general guidance on how to implement a cursor based pagination system for a database. Some general guidance:
> 
> * Read this [https://medium.com/swlh/how-to-implement-cursor-pagination-like-a-pro-513140b65f32](this) article on what queries you would expect from a cursor based pagination system
> * Rethink the returned value of your API (e.g. it is the page itself, plus a prev_ and next_ value)
> * Make sure you have a unique, sorted column in your database (this is a requirement!)
> * Rethink the parameter(s) that your api receives; does it receive only 1 param in the format of `next__{id}` or `prev__{id}` or does it have multiple?
> * If you have very specific questions on problems that you run into, you might be better helped on Stackoverflow.
> 
> In general, cursor based pagination can greatly speed up **certain** use cases, but offset pagination is absolutely not dead.
> 
> Good

Thank you very much for your reply. That offset or cursor pagination stuff is really an opinionated subject

**Comment by shellking4** [comment]
> We use the offset based pagination:
> 
> https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/490c554e23343eec0736b06e59b2108fdd057fdc/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/crud/base.py#L29
> 
> Satisfied about performance etc.. (even extended it with filters/search)

Thank you very much sir. I think I'll go for the offset one for my current project. Thank you for your reply

**Comment by acidjunk** [comment]
U R Welcome: you might like this even more: https://github.com/acidjunk/fastapi-postgres-boilerplate/blob/master/server/crud/base.py

**Comment by shellking4** [comment]
> U R Welcome: you might like this even more: https://github.com/acidjunk/fastapi-postgres-boilerplate/blob/master/server/crud/base.py

Yeah very helpful sir it gives me more insight on the implementation. Thank you sir
